src/OptimalControlUtilities/include/BipedalLocomotionControllers/OptimalControlUtilities/ControlProblemElements.h

```diff
@@ -0,0 +1,745 @@
+     * @param kinDyn an iDynTree kinDyn computation object
+     * @param handler the variable handler object
+     * @param framesInContact vector containing the frames in contact. Each element of the vector is
+     * a pair containing the name of the frame in the variableHandler and in the model.
+     * @param regularizationMatrix joints regularization mass matrix. It can be used to handle
+     * the motor reflected inertia
+     * @throw std::runtime_error if the frame is not defined or the regularizationMatrix size is
+     * no coherent with the number of joints
+     */
+    SystemDynamicsElement(iDynTree::KinDynComputations& kinDyn,
+                          const VariableHandler& handler,
+                          const std::vector<std::pair<std::string, std::string>>& framesInContact,
+                          const iDynTree::MatrixDynSize& regularizationMatrix);
+
+    /**
+     * Get (and compute) the element matrix
+     * @return the element matrix
+     */
+    virtual const iDynTree::MatrixDynSize& getA() final;
+
+    /**
+     * Get (and compute) the element vector
+     * @return the element vector
+     */
+    virtual const iDynTree::VectorDynSize& getB() final;
+};
+
+/**
+ * CentroidalLinearMomentumElement describes the Centroidal linear momentum
+ */
+class CentroidalLinearMomentumElement : public CostFunctionOrEqualityConstraintElement
+{
+private:
+    iDynTree::Vector3 m_VRP; /**< Position of the Virtual Repellent Point */
+    double m_robotMass; /**< Mass of the robot */
+
+public:
+    /**
+     * Constructor.
+     * @param kinDyn an iDynTree kinDyn computation object
+     * @param handler the variable handler object
+     * @param framesInContact vector containing the frames in contact. Each element of the vector is
+     * the name of the frame in the variableHandler
+     * @throw std::runtime_error if the frame is not defined
+     */
+    CentroidalLinearMomentumElement(iDynTree::KinDynComputations& kinDyn,
+                                    const VariableHandler& handler,
+                                    const std::vector<std::string>& framesInContact);
+
+    /**
+     * Set the desired VRP
+     * @param VRP position of the virtual repellent point
+     */
+    void setVRP(const iDynTree::Vector3& VRP);
+
+    /**
+     * Get (and compute) the element matrix
+     * @return the element matrix
+     */
+    virtual const iDynTree::MatrixDynSize& getA() final;
+
+    /**
+     * Get (and compute) the element vector
+     * @return the element vector
+     */
+    virtual const iDynTree::VectorDynSize& getB() final;
+};
+
+/**
+ * CentroidalAngularMomentumElement describes the Centroidal angular momentum
+ */
+class CentroidalAngularMomentumElement : public CostFunctionOrEqualityConstraintElement
+{
+private:
+    std::unique_ptr<PositionPID> m_pid; /**< Linear PID */
+    std::vector<Frame> m_framesInContact; /**< Vectors containing the frames in contact with the
```

Comment:
The same for the linear part. This element has a generic name but it is already specific to the final controller you are considering.